File: python/koans/06_reusable_barrier/test_reusable_barrier.py
import time

from koan_utils import EventLog, ThreadRunner, jitter

from reusable_barrier import ReusableBarrier


def assert_rounds_atomic(log, n, rounds):
    """Per round r: all n arrivals precede every departure of round r, and
    nobody arrives at round r+1 before everyone has arrived at round r."""
    for r in range(rounds):
        arrive, depart = f"arrive:{r}", f"depart:{r}"
        assert log.count(arrive) == n, (
            f"round {r}: expected {n} arrivals, saw {log.count(arrive)}"
        )
        assert log.count(depart) == n, (
            f"round {r}: expected {n} departures, saw {log.count(depart)}"
        )
        log.assert_before(arrive, depart)
        if r + 1 < rounds:
            log.assert_before(arrive, f"arrive:{r + 1}")


def run_rounds(n, rounds, join_timeout=15):
    barrier = ReusableBarrier(n)
    log = EventLog()
    runner = ThreadRunner()

    def worker():
        for r in range(rounds):
            jitter()
            log.record(f"arrive:{r}")
            barrier.wait()
            log.record(f"depart:{r}")
            jitter()

    for _ in range(n):
        runner.spawn(worker)
    runner.join_all(timeout=join_timeout)
    return log


def test_rounds_are_atomic():
    n, rounds = 4, 8
    log = run_rounds(n, rounds)
    assert_rounds_atomic(log, n, rounds)


def test_nobody_leaks_through_early():
    n = 4
    barrier = ReusableBarrier(n)
    departed = EventLog()
    runner = ThreadRunner()

    def worker(r):
        barrier.wait()
        departed.record(f"depart:{r}")

    for r in range(2):  # two consecutive rounds of the same barrier object
        for _ in range(n - 1):
            runner.spawn(worker, r)
        time.sleep(0.25)  # plenty of time to misbehave
        runner.raise_worker_errors()
        assert departed.count(f"depart:{r}") == 0, (
            f"round {r}: {departed.count(f'depart:{r}')} thread(s) got past "
            f"the barrier with only {n - 1} of {n} arrived"
        )
        runner.spawn(worker, r)  # the nth arrival springs the door
        runner.join_all(timeout=5)
        assert departed.count(f"depart:{r}") == n


def test_many_rounds_stress():
    # Uses the split-phase API (work between phase1 and phase2 is allowed),
    # which is where a resettable-in-name-only barrier lets a fast thread
    # lap the field.
    n, rounds = 6, 40
    barrier = ReusableBarrier(n)
    log = EventLog()
    runner = ThreadRunner()

    def worker():
        for r in range(rounds):
            jitter(1.0)
            log.record(f"arrive:{r}")
            barrier.phase1()
            jitter(3.0)  # some "work" at the critical point between phases
            barrier.phase2()
            log.record(f"depart:{r}")

    for _ in range(n):
        runner.spawn(worker)
    runner.join_all(timeout=15)
    total = sum(log.count(f"depart:{r}") for r in range(rounds))
    assert total == n * rounds, f"expected {n * rounds} departures, saw {total}"
    assert_rounds_atomic(log, n, rounds)


def test_works_for_two_batches_of_waits():
    n = 3
    barrier = ReusableBarrier(n)
    for _ in range(2):  # batch of n waits, plain join, then batch again
        runner = ThreadRunner()
        for _ in range(n):
            runner.spawn(barrier.wait)
        runner.join_all(timeout=5)

File: cpp/koans/06_reusable_barrier/test_reusable_barrier.cpp
#include "koan_test.hpp"
#include "reusable_barrier.hpp"

#include <memory>
#include <thread>

using namespace koans;

namespace {

// Per round r: all n arrivals precede every departure of round r, and
// nobody arrives at round r+1 before everyone has arrived at round r.
void assert_rounds_atomic(const EventLog& log, int n, int rounds) {
    for (int r = 0; r < rounds; ++r) {
        auto arrive = "arrive:" + std::to_string(r);
        auto depart = "depart:" + std::to_string(r);
        KOAN_ASSERT_MSG(log.count(arrive) == static_cast<std::size_t>(n),
                        "round " + std::to_string(r) + ": expected " +
                            std::to_string(n) + " arrivals, saw " +
                            std::to_string(log.count(arrive)));
        KOAN_ASSERT_MSG(log.count(depart) == static_cast<std::size_t>(n),
                        "round " + std::to_string(r) + ": expected " +
                            std::to_string(n) + " departures, saw " +
                            std::to_string(log.count(depart)));
        log.assert_before(arrive, depart);
        if (r + 1 < rounds)
            log.assert_before(arrive, "arrive:" + std::to_string(r + 1));
    }
}

EventLog run_rounds(int n, int rounds, std::chrono::milliseconds join_timeout) {
    auto barrier = std::make_shared<ReusableBarrier>(n);
    auto log = std::make_shared<EventLog>();
    ThreadRunner runner;
    for (int t = 0; t < n; ++t) {
        runner.spawn([barrier, log, rounds] {
            for (int r = 0; r < rounds; ++r) {
                jitter();
                log->record("arrive:" + std::to_string(r));
                barrier->wait();
                log->record("depart:" + std::to_string(r));
                jitter();
            }
        });
    }
    runner.join_all(join_timeout);
    return *log;
}

}  // namespace

KOAN_TEST(rounds_are_atomic) {
    constexpr int n = 4, rounds = 8;
    auto log = run_rounds(n, rounds, 15000ms);
    assert_rounds_atomic(log, n, rounds);
}

KOAN_TEST(nobody_leaks_through_early) {
    constexpr int n = 4;
    auto barrier = std::make_shared<ReusableBarrier>(n);
    auto departed = std::make_shared<EventLog>();
    ThreadRunner runner;
    for (int r = 0; r < 2; ++r) {  // two consecutive rounds, same barrier
        auto tag = "depart:" + std::to_string(r);
        for (int i = 0; i < n - 1; ++i) {
            runner.spawn([barrier, departed, tag] {
                barrier->wait();
                departed->record(tag);
            });
        }
        std::this_thread::sleep_for(250ms);  // plenty of time to misbehave
        KOAN_ASSERT_MSG(departed->count(tag) == 0,
                        "round " + std::to_string(r) + ": " +
                            std::to_string(departed->count(tag)) +
                            " thread(s) got past the barrier with only " +
                            std::to_string(n - 1) + " of " + std::to_string(n) +
                            " arrived");
        runner.spawn([barrier, departed, tag] {  // the nth springs the door
            barrier->wait();
            departed->record(tag);
        });
        runner.join_all(5000ms);
        KOAN_ASSERT_EQ(departed->count(tag), static_cast<std::size_t>(n));
    }
}

// Uses the split-phase API (work between phase1 and phase2 is allowed),
// which is where a resettable-in-name-only barrier lets a fast thread lap
// the field.
KOAN_TEST(many_rounds_stress) {
    constexpr int n = 6, rounds = 40;
    auto barrier = std::make_shared<ReusableBarrier>(n);
    auto log = std::make_shared<EventLog>();
    ThreadRunner runner;
    for (int t = 0; t < n; ++t) {
        runner.spawn([barrier, log] {
            for (int r = 0; r < rounds; ++r) {
                jitter(1);
                log->record("arrive:" + std::to_string(r));
                barrier->phase1();
                jitter(3);  // "work" at the critical point between phases
                barrier->phase2();
                log->record("depart:" + std::to_string(r));
            }
        });
    }
    runner.join_all(15000ms);
    std::size_t total = 0;
    for (int r = 0; r < rounds; ++r)
        total += log->count("depart:" + std::to_string(r));
    KOAN_ASSERT_EQ(total, static_cast<std::size_t>(n * rounds));
    assert_rounds_atomic(*log, n, rounds);
}

KOAN_TEST(works_for_two_batches_of_waits) {
    constexpr int n = 3;
    auto barrier = std::make_shared<ReusableBarrier>(n);
    for (int batch = 0; batch < 2; ++batch) {  // batch, plain join, batch again
        ThreadRunner runner;
        for (int i = 0; i < n; ++i)
            runner.spawn([barrier] { barrier->wait(); });
        runner.join_all(5000ms);
    }
}
